File: algorithms/268.Missing_Number/Missing_Number.cc
<<<<<<< HEAD
class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int result = nums.size();
        int i=0;

        for(int num:nums){
            result ^= num;
            result ^= i;
            i++;
        }

        return result;
    }
};

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int ret = (nums.size() + 1) * nums.size() /2;
        for(auto i: nums)
            ret -= i;
        return ret;
    }
};
=======
class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int result = nums.size();
        int i=0;

        for(int num:nums){
            result ^= num;
            result ^= i;
            i++;
        }

        return result;
    }
};

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int ret = (nums.size() + 1) * nums.size() /2;
        for(auto i: nums)
            ret -= i;
        return ret;
    }
};
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74

File: algorithms/268.Missing_Number/Missing_Number.java
public class Solution {
    public int missingNumber(int[] nums) {
        int sum = 0;
        for(int num: nums)
            sum += num;

        return (nums.length * (nums.length + 1) )/ 2 - sum;
    }
}
public class Solution {
    public int missingNumber(int[] nums) {
    int sum = 0;
    for (int i = 0; i <= nums.length; ++i) sum ^= i;
    for (int i = 0; i < nums.length; ++i) sum ^= nums[i];
    return sum;
}
}
1.XOR

public int missingNumber(int[] nums) { //xor
    int res = nums.length;
    for(int i=0; i<nums.length; i++){
        res ^= i;
        res ^= nums[i];
    }
    return res;
}

2.SUM

public int missingNumber(int[] nums) { //sum
    int len = nums.length;
    int sum = (0+len)*(len+1)/2;
    for(int i=0; i<len; i++)
        sum-=nums[i];
    return sum;
}

3.Binary Search

public int missingNumber(int[] nums) { //binary search
    Arrays.sort(nums);
    int left = 0, right = nums.length, mid= (left + right)/2;
    while(left<right){
        mid = (left + right)/2;
        if(nums[mid]>mid) right = mid;
        else left = mid+1;
    }
    return left;
}

Summary:

If the array is in order, I prefer Binary Search method. Otherwise, the XOR method is better.